day7/day_test.go

```diff
@@ -0,0 +1,53 @@
+package day7_test
+
+import (
+	"fmt"
+	. "ryepup/advent2021/day7"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestPart1(t *testing.T) {
+	testCases := map[string]int{"test.txt": 37, "input.txt": 345197}
+	for path, expected := range testCases {
+		t.Run(path, func(t *testing.T) {
+			if expected == -1 {
+				t.Skip("TODO: provide an expected value")
+			}
+			n, err := Part1(path)
+			require.Nil(t, err)
+			require.Equal(t, expected, n)
+		})
+	}
+}
+
+func TestPart2(t *testing.T) {
+	testCases := map[string]int{"test.txt": 168, "input.txt": 96361606}
+	for path, expected := range testCases {
+		t.Run(path, func(t *testing.T) {
+			if expected == -1 {
+				t.Skip("TODO: provide an expected value")
+			}
+			n, err := Part2(path)
+			require.Nil(t, err)
+			require.Equal(t, expected, n)
+		})
+	}
+}
+
+func BenchmarkPart2(b *testing.B) {
```

Comment:
Results:

```text
goos: linux
goarch: amd64
pkg: ryepup/advent2021/day7
cpu: Intel(R) Core(TM) i7-10875H CPU @ 2.30GHz
BenchmarkPart2/P:0_C:0-16         	       3	 460832852 ns/op	   49288 B/op	       9 allocs/op
BenchmarkPart2/P:0_C:1-16         	      19	  60033013 ns/op	  222946 B/op	     112 allocs/op
BenchmarkPart2/P:0_C:2-16         	      20	  55571215 ns/op	  222885 B/op	     113 allocs/op
BenchmarkPart2/P:0_C:3-16         	      26	  45287272 ns/op	  222655 B/op	     110 allocs/op
BenchmarkPart2/P:1_C:0-16         	      22	  52479210 ns/op	  260651 B/op	    3816 allocs/op
BenchmarkPart2/P:1_C:1-16         	       7	 143178662 ns/op	  455013 B/op	    4261 allocs/op
BenchmarkPart2/P:1_C:2-16         	      10	 106617758 ns/op	  436376 B/op	    4062 allocs/op
--- FAIL: BenchmarkPart2/P:1_C:3
    /workspaces/advent-of-code-2021/day7/day_test.go:47: invalid strategy combination
```

Reply:
I could get prettier names, but I didn't want to figure out `stringer` and `go generate` tonight.

As expected memoizing saves a lot of time at the expense of memory, and parallelism saves time as long as we're not having to coordinate.

The leaderboard:

1. For loop with non-locking memoized fuel cost
2. Goroutines with non-memoized fuel cost - burn cpu on redundant computation on all cores
3. For loop with RWMutex memoized fuel cost - we have many more reads than writes, so that indicated `RLock` is cheaper than `Lock`
4. For loop with Mutex memoized fuel cost
5. Goroutines with RWMutex memoized fuel cost - the synchronization cost more than the processing it's guarding
6. Goroutines with Mutex memoized fuel cost
7. For loop with non-memoized fuel cost - burn cpu but it's serialized to one core